src/profile-logic/zip-files.js

```diff
@@ -0,0 +1,269 @@
+        j + 1 === pathParts.length ? zipEntries.files[fullPath] : null;
+      pathToFilesTableIndex.set(path, index);
+      // Remember this index as the prefix.
+      prefixIndex = index;
+    }
+  }
+  return filesTable;
+}
+
+export function getZipFileMaxDepth(zipFileTable: ZipFileTable | null): number {
+  if (!zipFileTable) {
+    return 0;
+  }
+  let maxDepth = 0;
+  for (let i = 0; i < zipFileTable.length; i++) {
+    maxDepth = Math.max(maxDepth, zipFileTable.depth[i]);
+  }
+  return maxDepth;
+}
+
+export class ZipFileTree {
+  _zipFileTable: ZipFileTable;
+  _parentToChildren: null | Map<
+    IndexIntoZipFileTable | null,
+    IndexIntoZipFileTable[]
+  >;
+  _displayDataByIndex: Map<IndexIntoZipFileTable, ZipDisplayData>;
+  _zipFileUrl: string;
+
+  constructor(zipFileTable: ZipFileTable, zipFileUrl: string) {
+    this._zipFileTable = zipFileTable;
+    this._zipFileUrl = zipFileUrl;
+    this._displayDataByIndex = new Map();
+  }
+
+  getRoots(): IndexIntoZipFileTable[] {
+    return this.getChildren(null);
+  }
+
+  getChildren(
+    zipTableIndex: IndexIntoZipFileTable | null
+  ): IndexIntoZipFileTable[] {
+    const parentToChildMap = this._getParentToChildMap();
+    const children = parentToChildMap.get(zipTableIndex);
+    if (!children) {
+      throw new Error(
+        'Attempted to fetch the children from an unknown IndexIntoZipFileTable.'
+      );
+    }
+    return children;
+  }
+
+  _computeChildrenArray(
+    parentIndex: IndexIntoZipFileTable | null
+  ): IndexIntoZipFileTable[] {
+    const children = [];
+    for (
+      let childIndex = 0;
+      childIndex < this._zipFileTable.length;
+      childIndex++
+    ) {
+      if (this._zipFileTable.prefix[childIndex] === parentIndex) {
+        children.push(childIndex);
+      }
+    }
+    return children;
+  }
+
+  /**
+   * Create a Map of the parents to the children to make it O(1) to dynamically compute
+   * any property about the tree.
+   */
+  _getParentToChildMap(): Map<
+    IndexIntoZipFileTable | null,
+    IndexIntoZipFileTable[]
+  > {
+    let parentToChildren = this._parentToChildren;
+    if (!parentToChildren) {
+      parentToChildren = new Map();
+      parentToChildren.set(null, this._computeChildrenArray(null));
+
+      for (
+        let parentIndex = 0;
+        parentIndex < this._zipFileTable.length;
+        parentIndex++
+      ) {
+        const children = this._computeChildrenArray(parentIndex);
+        parentToChildren.set(parentIndex, children);
+      }
+    }
+    return parentToChildren;
+  }
+
+  hasChildren(zipTableIndex: IndexIntoZipFileTable): boolean {
+    return this.getChildren(zipTableIndex).length > 0;
+  }
+
+  getAllDescendants(
+    zipTableIndex: IndexIntoZipFileTable
+  ): Set<IndexIntoZipFileTable> {
+    const result = new Set([]);
+    for (const child of this.getChildren(zipTableIndex)) {
+      result.add(child);
+      for (const descendant of this.getAllDescendants(child)) {
```

Comment:
I also don't really like this loop as we'll create a new Set object for each folder; I think it would be better to pass the set along. Although this is probably not a big deal for a zip file, this is certainly an issue with the call tree... I'll look at it in my other PR.

Reply:
Keeping it as is for now.